perf(Terrain Variation): feature rework, miplevel & sample optimisation

davo0411 opened the pull request:
Just a draft for now so people can easier look at my code.

Some help with optimisation would be appreciated. Any attempts I have made to successfully reduce sample count has done little to nothing.

I will look further into optimising before nexus / when I can.

Outperforms dev by around 5%

The +1000 -600 should change dramatically between now and final. Lots to cut from EMAT before it is done.

Changes from dev:
- Fix for all mipmap issues or lines in terrain. Now uses worldspace uv not recomputed uv.
- Major rewrite of the lighting.hlsl terrain layer system to better respect the old version. Also better respects PBR implementations, the current dev math is just bad. 
- Full rewrite of the Terrain Variation function. Added a luminance-respecting blend operator to keep the details in the terrain better. Looks functionally indistinguishable from disabled up close now. 
- Rewrites of some extendedmaterials segments, nothing major. Added overloads for Terrain variation toggled off. This is the biggest thing that will change from now to final - this is redundant. But when i remove it and go back to just if def (tv), i get more weird mipmap issues. Will fix later before final.
- Added LOD Terrain tiling fix
- No distance factor, terrain is either offset or not. New system for offsets ensure the terrain detail is perfectly preserved.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Added non-stochastic terrain height and parallax mapping options for improved flexibility when tiling fix is disabled.
  - Introduced new methods for resetting terrain variation settings to defaults.

- **Improvements**
  - Simplified and unified terrain stochastic sampling for more consistent and efficient rendering.
  - Streamlined terrain variation settings UI, focusing on essential toggles and clearer tooltips.
  - Enhanced shader performance by removing distance-based blending and advanced runtime checks.

- **Bug Fixes**
  - Improved handling of terrain tiling fix and LOD terrain support.

- **Chores**
  - Updated internal parameter names and removed unused fields for better maintainability.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

davo0411 committed:
Terrain Variation Lite

davo0411 committed:
Always on Height Operator build 1

davo0411 committed:
Update TerrainVariation.hlsli

davo0411 committed:
Update TerrainVariation.h

davo0411 committed:
Update 2, working.

davo0411 committed:
update 3

davo0411 committed:
comment clean 1

davo0411 committed:
Merge remote-tracking branch 'upstream/dev' into TV-NO-DIST-ALWAYSON

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
LOD v1

davo0411 committed:
fix for mipmap issues & not compiling

davo0411 committed:
Merge branch 'TV-NO-DIST-ALWAYSON' of https://github.com/davo0411/skyrim-community-shaders into TV-NO-DIST-ALWAYSON

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
Fixed all bugs, comment cleanup

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
Worldspace based

davo0411 committed:
Fix

davo0411 committed:
Update Lighting.hlsl

davo0411 committed:
Merge branch 'TV-NO-DIST-ALWAYSON' of https://github.com/davo0411/skyrim-community-shaders into TV-NO-DIST-ALWAYSON

davo0411 committed:
Update TerrainVariation.hlsli

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
Lod optimisations 1

davo0411 committed:
Merge branch 'TV-NO-DIST-ALWAYSON' of https://github.com/davo0411/skyrim-community-shaders into TV-NO-DIST-ALWAYSON

davo0411 committed:
Big fix, uses texcoord0 now.

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
Mipmap fixes

davo0411 committed:
style: 🎨 apply clang-format changes

davo0411 committed:
Merge branch 'doodlum:dev' into TV-NO-DIST-ALWAYSON

davo0411 committed:
LOD Optimisation

davo0411 committed:
Mipmap fixes

davo0411 committed:
Merge branch 'TV-NO-DIST-ALWAYSON' of https://github.com/davo0411/skyrim-community-shaders into TV-NO-DIST-ALWAYSON

davo0411 committed:
tidy

davo0411 committed:
fix

davo0411 committed:
fix for dumb mistakes.

davo0411 commented:
@coderabbitai review

davo0411 commented:
@coderabbitai suggest improvements to the sample gathering system, and potential ways that it can be improved to reduce high sample count and misses shader cache events when close pixels are calculating very different uvs.

alandtse commented:
Please update the title to be more descriptive of what you're optimizing. Also consider reading the spec we use.
https://www.conventionalcommits.org/en/v1.0.0/#summary

alandtse commented:
Actually, I didn't see the full extent of files touched. This likely needs to be broken out into logically distinct PRs when ready. You will have some new features for TV, but you're also refactoring or improving performance of other sections that are not related at all to TV. Please plan appropriately.

davo0411 commented:
> Please update the title to be more descriptive of what you're optimizing. Also consider reading the spec we use. https://www.conventionalcommits.org/en/v1.0.0/#summary

Its a pretty big change but i hope the new title suffices.

alandtse commented:
> > Please update the title to be more descriptive of what you're optimizing. Also consider reading the spec we use. [conventionalcommits.org/en/v1.0.0#summary](https://www.conventionalcommits.org/en/v1.0.0/#summary)
> 
> bruh i didn't name it that, i named it Feat: Terrain Variation Major Rework

Yes, it was a bad title. Major rework is useless. I started to fix it based off what you had, but needed more info. However, given how big this is, that is why the title was crap. Please plan to break this up once it's out of draft PR as outlined in my prior comment.

davo0411 commented:
> Actually, I didn't see the full extent of files touched. This likely needs to be broken out into logically distinct PRs when ready. You will have some new features for TV, but you're also refactoring or improving performance of other sections that are not related at all to TV. Please plan appropriately.

Yes, as i said in the description the extent of changes isn't fully representative of the final pr (I have a lot to cut from Extended mats). Its mainly up so I can get some help with the feature as it stands. Only touched on 3 files, and all changes are related to tv. Its basically redoing what I had done in the original version better.

davo0411 committed:
accidental push

davo0411 committed:
Update TerrainVariation.cpp

davo0411 commented:
> You can ask an AI on how to break it up into manageable independent PRs. Some may need to be staged after other ones, but this is make the commit history clean and cherry-pickable. Giant omnibus PRs are hard to review and if they break stuff, may not be salvageable beyond a revert. Again, I haven't read all your changes because GitHub thinks the diffs are huge.
> 
> If you want a quick idea, look at how the AI broke up the summary into bullet points. In theory each could be it's own commit. You can probably figure out the minimum changes to do that. Again, this is when it's out of draft. Feel free to rewrite everything in draft if you want. I'm just letting you know what the PR review will be like. You're already sitting on 3-4 PRs and you have seen when they get in easy and when they take longer (particularly if other people are PRs too).

I understand that, and will work on smaller prs in future. However here, every single call to tv that i added in the original commit is changed as it is no longer passing the same parameters. I am no longer using the distance parameter. It is changing this across the couple hundred lines I had added originally, that is why it appears unweildly, however it is really only changing one word, not the 500 lines it appears as. I can split the individual file changes across prs, but it'll mean that the shaders are fundamentally broken until all is merged/complete.

Again, this draft is mostly so i can get help with the code itself. Final code, pr, formatting, is all not finalised. 
I understand the pr should be split or smaller, this was made abundantly clear in tv v1. This is on the same scale as the original feature.

davo0411 committed:
fixes

pre-commit-ci[bot] committed:
style: 🎨 apply pre-commit.ci formatting

Automated formatting by clang-format, prettier, and other hooks.
See https://pre-commit.ci for details.

davo0411 committed:
Merge remote-tracking branch 'upstream/dev' into TV-NO-DIST-ALWAYSON

davo0411 committed:
Minor height changes, +.5ms.

davo0411 committed:
Distance blend, much better perf.

davo0411 committed:
Update TerrainVariation.hlsli

davo0411 committed:
Emat separated

davo0411 committed:
Update TerrainVariation.hlsli

pre-commit-ci[bot] committed:
style: 🎨 apply pre-commit.ci formatting

Automated formatting by clang-format, prettier, and other hooks.
See https://pre-commit.ci for details.

davo0411 committed:
undo to emat changes

davo0411 committed:
Update ExtendedMaterials.hlsli

davo0411 committed:
Updates

davo0411 committed:
update

davo0411 committed:
EMAT Refactor part 1

davo0411 committed:
glint fix

davo0411 committed:
Swap to Miplevel over distance

davo0411 committed:
Miplevel final

davo0411 committed:
perf updates

davo0411 commented:
@coderabbitai give me an in depth plan of approach for refactoring Extendedmaterials.hlsli to reduce code repetition.

davo0411 committed:
Weight aware blend & sample gathering

davo0411 committed:
reintroduced miplevel fixes

davo0411 committed:
Quality & Performance Fix, no lines in terrain.

davo0411 committed:
Complex material fixes

davo0411 committed:
Merge branch 'dev' into TV-NO-DIST-ALWAYSON

pre-commit-ci[bot] committed:
style: 🎨 apply pre-commit.ci formatting

Automated formatting by clang-format, prettier, and other hooks.
See https://pre-commit.ci for details.

davo0411 commented:
@coderabbitai
How would you suggest refactoring ExtendedMaterials.hlsli, and would a system be possible to make the parallax system with terrain variation cheaper? Currently computing GetTerrainHeight on 3 samples is very expensive

davo0411 committed:
fix for parallaxsoftshadow
